scipy/optimize/_minimize.py

```diff
@@ -109,34 +109,37 @@ def minimize(fun, x0, args=(), method=None, jac=None, hess=None,
         obey any specified `bounds`.
     hess : {callable, '2-point', '3-point', 'cs', HessianUpdateStrategy}, optional
         Method for computing the Hessian matrix. Only for Newton-CG, dogleg,
-        trust-ncg, trust-krylov, trust-exact and trust-constr. If it is
-        callable, it should return the Hessian matrix:
+        trust-ncg, trust-krylov, trust-exact and trust-constr.
+        If it is callable, it should return the Hessian matrix:
 
             ``hess(x, *args) -> {LinearOperator, spmatrix, array}, (n, n)``
 
         where ``x`` is a (n,) ndarray and ``args`` is a tuple with the fixed
         parameters. LinearOperator and sparse matrix returns are only allowed
-        for 'trust-constr' method. Alternatively (not available for Newton-CG
-        or dogleg), the keywords {'2-point', '3-point', 'cs'} select a finite
-        difference scheme for numerical estimation. Or, objects implementing
-        the `HessianUpdateStrategy` interface can be used to approximate the
-        Hessian. Available quasi-Newton methods implementing this interface
-        are:
+        for the trust-constr method.
+        dogleg and trust-exact can only be used with a callable returning
+        an (n, n) array.
+        It is not mandatory to supply this keyword for trust-constr or
+        Newton-CG.
```

Comment:
I don't get this.

Reply:
Newton-CG and trust-constr can have `hess=None`. In the case of Newton-CG it's because it [approximates](https://github.com/scipy/scipy/blob/1f0cd312e1617df1631cf347619ac8cae58765be/scipy/optimize/_optimize.py#L1922) a `hessp` using [this function](https://github.com/scipy/scipy/blob/1f0cd312e1617df1631cf347619ac8cae58765be/scipy/optimize/_optimize.py#L1071).
The logic for Newton-CG is:

- if hess=None approximate a `hessp` using `approx_fhess_p` using finite-differences on the `jac`.
- if you have a `callable(hess)` use that.
- if you supply a `FD_METHOD` or a `HessianUpdateStrategy` then use that to approximate the Hessian via FD of the `jac`, then use that approximation in a `hessp` wrapper. This is different to the first option in that you can control how the FD occurs.
- if you have a callable(hessp) use that.

It's quite extended logic, and all the minimize methods are slightly different, which is why the docstring is tortuous. I think a table makes sense.

Reply:
ok thanks for the explanation!